ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSshell.cpp

```diff
@@ -90,21 +91,25 @@ string execute_shell_for_output(const string &command) {
         TranslateMessage(&msg);
         DispatchMessage(&msg);
       }
+      DWORD dwRead = 0;
+      DWORD dwAvail = 0;
+      PeekNamedPipe(hStdOutPipeRead, NULL, 0, NULL, &dwAvail, NULL);
+      if (dwAvail) {
+        char buffer[dwAvail];
+        BOOL success = ReadFile(hStdOutPipeRead, buffer, dwAvail, &dwRead, NULL);
+        if (success || dwRead) {
+          buffer[dwAvail] = 0;
+          output.append(buffer, dwAvail);
+        }
+      }
     }
     CloseHandle(hStdOutPipeWrite);
     CloseHandle(hStdInPipeRead);
-    char buffer[4096] = { };
-    DWORD dwRead = 0;
-    ok = ReadFile(hStdOutPipeRead, buffer, 4095, &dwRead, NULL);
-    while (ok == TRUE) {
-      buffer[dwRead] = 0;
-      ok = ReadFile(hStdOutPipeRead, buffer, 4095, &dwRead, NULL);
-    }
     CloseHandle(hStdOutPipeRead);
     CloseHandle(hStdInPipeWrite);
     CloseHandle(pi.hProcess);
     CloseHandle(pi.hThread);
-    return shorten(widen(buffer));
+    return output;
```

Comment:
What makes you think output is UTF-8 from this API?

Reply:
Josh told me calling shorten(widen(something)) returns nothing different from what something already would return by itself, and that if my code worked with utf8 (which it does in my testing) that would imply i can just return the buffer appended to the output variable directly instead of calling shorten and widen on it, which has proven to be redundant, as Josh suspected.

Reply:
Josh explained to me now as well that is does not make a difference. It is not standardized what format stdout happens to be in. This lies at the "boundary" of the program and therefore is not something we need to worry about.
http://utf8everywhere.org/

We could perhaps later provide shorten and widen in the EDL user land though. For now, there's nothing else to be said about this.